ECMA spec addendum dealing with static interface methods

trylek opened the pull request:
Opening this review to discuss ECMA spec amendments related to static interface methods.

trylek committed:
ECMA spec addendum dealing with static interface methods

jkotas commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+Some aspects that are unclear to me now:
+
+* Especially in light of the default interface work, would it make sense to allow "overriding" (implementing) a static interface method
```
I think this was covered in https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-02-24.md#static-abstract-members-in-interfaces: "In order to scope the initial implementation, this proposal intentionally limits such members to only abstract members, not virtual members"

davidwrighton replied:
Yes. Very specifically, support for default implementation is not allowed. If a static method is virtual, it must be abstract.

trylek replied:
Got it, thanks for the clarification.

davidwrighton commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+### II.12.2 Implementing virtual methods on interfaces
+
+For now I'm stuck in this complex section, I would welcome initial feedback as to whether what I've shared so far makes any sense.
```
In the section where it specifies that "The VES shall", replace the two uses of "public virtual" with "public virtual instance"

trylek replied:
Fixed in 2nd commit.

davidwrighton commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+(It is unclear to me what is the expected general behavior of generic static virtual methods; it should probably be stated here).
+
+### II.9.11, Constrains on Generic Parameters
```
There are two ways to handle this.

1. Require that even in the presence of an abstract type, the static interface methods must be fully implemented. (This would be a change to II.12.2, "If the current class is not abstract and..." would need qualification to specify that static interface methods must be implemented. This would cover the instantiation over class case, but there is another detail, which is that instantiated over the interface which defines the constraint which also need to be disabled. Thus I prefer option 2.
OR
2. We adjust the rules for constraints. Section II.9.11 is adjusted so that instead of specifying that a constraint requires an "assignable-to" relationship, that the relationship is "assignable-to" and *implements-static-interface-methods-of*. Where *implements-static-interface-methods-of* requires that the instantiating type satisfy the interface static methods.

If we go down option 2, we can also have the question of do we implement a type load check at all. My preference is to do both, where we have a check for concrete types at type load time, and the constraint check is also updated to cover the abstract class scenario as well as the interface scenario.

trylek replied:
I have adjusted section II.9.11 according to David's "2nd option" in 2nd commit.

trylek replied:
Thanks for clarifying, I'm closing this conversation for now as I believe we've reached a resolution. Please reopen it if you believe there's more to discuss.

davidwrighton commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+* Interface implementation examples (II.12.2.1, page 159)
+* Defining, referencing and calling methods (II.15, page 177)
+* constrained. - (prefix) invoke a member on a value of a variable type (III.2.1, page 316)
+* call - call a method (III.3.19, page 342)
```
Good point. I think we should strongly consider that. I'll talk to the language folks on that concern.

trylek replied:
Added to the list of sections that may need adjusting in 2nd commit.

trylek replied:
I have incorporated the envisioned ldftn changes in all sections I'm aware of.

davidwrighton reviewed (commented):
This is my first pass at this. I think we'll need a longer writeup for the IL opcode(s), but its a good first step here.

davidwrighton commented:
I've spoken to Mads and the rest of the static interfaces group, and there is general agreement that we'll need delegates and possibly function pointers.

davidwrighton commented:
Continuing suggestions

Continuing suggestions for spec changes.

[II.22.26] MethodDef
Replace 7.a "Static | Virtual" in the metadata validation rules with "Static | Virtual | Abstract"

Add a new metadata validation rule
41. If the owner of this method is not an interface, then if Flags.Static is 1 then Flags.Virtual must be 0.

[II.22.27] MethodImpl
Change metadata validation rule 7.
Replace "The method indexed by MethodBody shall be virtual" with "The method indexed by MethodBody shall be non-virtual if the method indexed by MethodDeclaration is static. Otherwise it shall be virtual."

[III.2.1] constrained prefix
Adjust stack transition diagram into a table. 

| Prefix and instruction pair                   | Stack Transition
| constrained. _thisType_ callvirt _method_     | ..., ptr, arg1, ... argN -> ..., ptr, arg1, ... argN
| constrained. _implementorType_ call _method_  | ..., arg1, ... argN -> ..., arg1, ... argN
| constrained. _implementorType_ ldftn _method_ | ..., ftn -> ..., ftn

Replace 

"The constrained. prefix is permitted only on a callvirt instruction. The" with 

"The constrained. prefix is permitted on the callvirt, call and ldftn instructions. 

When followed by the callvirt instruction, the"

Then at the end of the description add the following text.

"When followed by the call instruction or the ldftn instruction, _method_ must refer to a virtual static method defined on an interface. The behavior of the constrained. prefix is to change the method that the call or ldftn instruction refer to to be the method on _implementorType_ which implements the virtual method described by _method_ as defined in partition II, Introducing and overriding virtual methods.

In the correctness section replace "callvirt instruction" with "callvirt, call, or ldftn instruction"

In the verifiability section add "The _implementorType_ must be constrained to implement the interface that is the owning type of _method_. If the constrained. prefix is applied to a call or ldftn instruction, _method_ must be a virtual static method."

trylek commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+### II.9.10, Explicit Method Overrides
+
+(It is unclear to me what is the expected general behavior of generic static virtual methods; it should probably be stated here).
```
Added in 2nd commit, thanks for pointing that out.

trylek replied:
I have also removed the reference to section II.9.10 from the list of paragraphs to amend in the 2nd commit.

trylek replied:
As a non-native English speaker I may be also completely off but I'm wondering whether we should expand the change to the paragraph II.9.9 to something like *for instance virtual methods as opposed to static virtual methods* as I believe that the formulation *for instance* also means *for example, e.g.* and it may be confusing for the reader.

trylek replied:
Thinking about it some more, it's probably easiest to change the word ordering to what you're suggesting below, <code>for virtual instance methods</code>, that resolves the ambiguity I believe.

trylek replied:
Applied in the change, closing the conversation to streamline the review.

trylek commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+### II.10.3 Introducing and overriding virtual methods
+
+For now I don't see any particular formulation in this section that would need amending. We may want to mention virtual static
```
Fixed in 2nd commit.

trylek commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -366,3 +366,72 @@ In addition to the guarantees that apply to all type initializers, the CLI shall
+
+(Add to the end of the 1st paragraph)
+
+Interfaces may define static virtual methods that get resolved at runtime based on actual types involved.
```
Added in 2nd commit.

trylek commented:
@davidwrighton - I am somewhat confused about your comment

> [II.22.26] MethodDef
Replace 7.a "Static | Virtual" in the metadata validation rules with "Static | Virtual | Abstract"

7.a is "static | final" and 7.b is "static | virtual". Assuming you suggest to modify the section 7.b that seems to better match your quotation, does that mean that "static virtual abstract" methods are invalid? Please note that the section 7 refers to invalid flag combinations. I suspect we probably need something like <code>Static | Virtual | !Abstract</code> but I'm not sure how to express this.

trylek committed:
Address initial PR feedback; clarify most of the relevant sections

trylek commented:
@davidwrighton - I believe I have addressed all your PR feedback, can you please take another look when you have a chance? From the E2E perspective, I only have a bunch of open questions and I haven't yet populated the "examples" section, otherwise I believe I've covered all relevant sections that you and I found in the ECMA spec.

trylek committed:
Remove trailing whitespace triggering Markdownlint failures

trylek committed:
Attempt to fix the table

davidwrighton commented:
@trylek, I think this is close enough for us to start working on runtime test cases/implementation.

@AlekseyTs, I'd like you to take a look through the changes and see if they seem appropriate to you as well.

trylek committed:
Tiny fix + initial attempt at providing examples of static interface methods

trylek committed:
Initial self-review fixes, the woes of code that doesn't compile

trylek committed:
More fixes

trylek committed:
Yet more fixes

AlekseyTs commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,393 @@ The algorithm is amended as follows:
+(Add second paragraph)
+
+Static interface methods may be marked as virtual. Valid object types implementing such interfaces shall provide implementations
+for these methods by means of Method Implementations (II.15.1.4). Polymorphic behavior of calls to these methods is facilitated
```
I think virtual methods don't have to be implemented. They can be, but don't have to be.

AlekseyTs commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,393 @@ The algorithm is amended as follows:
+
+(Add second paragraph)
+
+Static interface methods may be marked as virtual. Valid object types implementing such interfaces shall provide implementations
```
Are we going to allow abstract static members?

trylek replied:
I believe that for now we're only planning abstract static methods i.e. the interface cannot provide a "default implementation" for the methods. As JanK wrote in another comment above:

*I think this was covered in https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-02-24.md#static-abstract-members-in-interfaces: "In order to scope the initial implementation, this proposal intentionally limits such members to only abstract members, not virtual members"*

and DavidWr further clarified that

*Yes. Very specifically, support for default implementation is not allowed. If a static method is virtual, it must be abstract.*

AlekseyTs commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,393 @@ The algorithm is amended as follows:
+
+The first *TypeSpec::MethodName* pair specifies the virtual method that is being overridden, and shall
+be either an inherited virtual method or a virtual method on an interface that the current type
+implements. The remaining information specifies the virtual instance or non-virtual static method that
```
Should we explicitly disallow mismatch of instance vs. static in the pair?

trylek replied:
The existing ECMA spec says in the .override section (II.10.3.2, page 147):

*The .override directive specifies that a virtual method shall be implemented (overridden), in this type,
by a virtual method with a different name, but with the same signature. This directive can be used to
provide an implementation for a virtual method inherited from a base class, or a virtual method
specified in an interface implemented by this type.*

As you see, static may be used to override an instance method (there's no harm in dropping the this pointer when not needed), of course not the other way round, that may be worth clarifying.

trylek replied:
Sorry, I looked at the updated comment, the original ECMA spec doesn't allow overriding virtuals with statics, I need to double-check how it got to the update - I think it was meant to clarify the distinction between virtual instance and virtual static methods but I think it's quite confusing the way it stands now, I'll fix that.

trylek replied:
Updated in 12th commit - @davidwrighton, can you please take a look at the wording change I made in that commit and let me know if you find that reasonable?

davidwrighton replied:
@AlekseyTs, we already explicitly disallow mismatch of instance vs static as the signature in Ecma335 includes the detail of whether or not a method is instance vs static. I don't like the additional change @trylek added in change 12 which is redundant. I would prefer to revert back to the wording in change 11.

AlekseyTs replied:
@davidwrighton 

>  we already explicitly disallow mismatch of instance vs static as the signature in Ecma335 includes the detail of whether or not a method is instance vs static.

Could you point me to the relevant section of provide a quote from it.

trylek replied:
I can easily revert the 12th commit but in light of your clarifications I'm wondering what was the purpose of adding the mention of non-virtual statics to the ECMA spec in the first place?

*[DavidWr] In section 10.3.2, replace ... "by a virtual method with a different name" with "by a virtual instance method with a different name, **or a non-virtual static method**"*

davidwrighton replied:
@AlekseyTs The definition of a method signature is in I.8.6.1.5, and the quote is "A method signature is composed of 
 a calling convention, 
 the number of generic parameters, if the method is generic, "

Also, II.15.3 where there is a definition of calling convention which describes how to specify an instance method (as opposed to a static method.)

The spec is somewhat inconsistent in defining whether or not a calling convention includes the static vs instance detail as in partition I.8.6.1.6 it calls out calling conventions and static vs instance separately. But even there, a matching signature is considered to include details such as if it is instance vs static.

@trylek The point of my wording is that we are not intending to allow virtual static methods to exist on classes. Thus any implementation itself cannot be virtual. The change in wording is to allow a non-virtual static method to implement a virtual static interface method. This mixing of virtual and non-virtual is not permitted for interface methods (even though C# allows such logic. C# implements such methods by silently making those methods virtual).

trylek commented:
@davidwrighton - I have added an initial attempt at static interface method examples; I apologize to the C# team for tons of bugs I have certainly made in the syntax, writing valid code that doesn't compile is hard. Cc @tannergooding as the first examples I tried to provide are mostly based on the polymorphic math. I'll welcome any feedback as to how to steer this section of the proposed spec augment.

trylek committed:
Removed superfluous type DoublePair; more syntax fixes

AlekseyTs commented:
Should we reinstate variance safety requirement for signatures of **virtual** static interface methods? Recently we removed that requirement for static methods since the couldn't be virtual and were not subject to variance.

trylek committed:
More syntax fixes :-)

trylek committed:
Clarify naming as the local actually represents the inverse factorial

trylek committed:
Clarify the .override directive per Aleksey's PR feedback

davidwrighton commented:
@AlekseyTs I'm not familiar with the details of the variance safety work. Could you provide a reference? At first blush I don't think we need any adjustment, as this form of variant dispatch is not effected by the instance type of one of the parameters, but I'd need more details to respond with confidence.

davidwrighton commented:
@trylek With respect to the static method examples, I believe they are roughly correct and good enough for now, but we'll need to adjust them to be proper C# once the design is finalized. As compared to the BCL design that @tannergooding is working on, I'm not sure that we need them to match up at all. although we might want the example to use the curiously recurring generic pattern.

AlekseyTs commented:
@davidwrighton 

> I'm not familiar with the details of the variance safety work. Could you provide a reference? At first blush I don't think we need any adjustment, as this form of variant dispatch is not effected by the instance type of one of the parameters, but I'd need more details to respond with confidence.

See https://github.com/dotnet/runtime/pull/40152. It feels like the "instance type" (not a parameter type) can actually implement variant "compatible" interface and since virtual static methods now a subject for a virtual dispatch, variance safety rules should be enforced for them.

davidwrighton commented:
@AlekseyTs  Good catch. Yes, we need to adjust II.9.7 and replace "There are no restrictions on static members, instance constructors, or ..." with "There are no restrictions on *non-virtual* static members, instance constructors, or ..."

trylek committed:
Address additional David's and Aleksey's PR feedback

I have added the change to II.9.7 per Aleksey's and David's
suggestion. I have reverted the changed wording in II.10.3.2 per
David's feedback.

Thanks

Tomas

AlekseyTs commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,389 @@ The algorithm is amended as follows:
+A generic parameter declared on a generic class or generic method can be *constrained* by one or more
+type (for encoding, see *GenericParamConstraint* table in paragraph II.22.21) and by one or more special
+constraints (paragraph II.10.1.7). Generic parameters can be instantiated only with generic arguments that are
+*assignable-to* (paragraph I.8.7.3) (when boxed) and *implements-all-static-interface-methods-of* (**paragraph
```
Do we want to use wording used in proposed language specification at https://github.com/dotnet/csharplang/blob/main/proposals/statics-in-interfaces.md#interface-constraints-with-static-abstract-members? Specifically: "When `I` has static abstract members this needs to be further restricted so that `T` cannot itself be an interface."

davidwrighton replied:
That would be an additional restriction on top of this one. This is designed to cover the case where abstract classes can be used as generic parameters.

AlekseyTs replied:
>That would be an additional restriction on top of this one.

Could you elaborate please?

> This is designed to cover the case where abstract classes can be used as generic parameters.

Abstract classes are not interfaces. Right?

davidwrighton replied:
So, we need to restrict the generic to fully implement the constraint. Normally, this is done by requiring that the type fully implement the interface through one of two mechanisms.
1. The type must be a reference type, in which case any actual object instance must be a non-abstract type and therefore have a full implementation of the interfaces it and its base types claim to implement.
2. The type must be a valuetype which must by definition cannot be abstract type, and thus satisfy the rule above.

With the addition of the static interfaces feature, instantiations over abstract classes bring along the possibility that the type may not fully implement the interface. Thus we need to protect against that.

Now, as you note, the C# language proposal suggests that we simply prevent any interface from being used as a generic for a static interface. I would prefer to word this restriction in a different way. My preference would be to requite that the type which is used as the generic parameter to provide a full set of implementations, and to then restrict the ability to use a MethodImpl on an interface type to work with these static methods.

trylek committed:
Clarify that the MethodImpls must point to the same class, not base

AlekseyTs commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -502,8 +502,8 @@ implements. The remaining information specifies the virtual instance or non-virt
-method to be used to provide an implementation, a conforming program shall provide a virtual instance method
-actually implemented directly on the type or a static method implemented on the type or on a base type.
+method to be used to provide an implementation, a conforming program shall provide a virtual instance
+or static method actually implemented directly on the type.
```
Is there anything to change in MethodImpl section? I think we also agreed to specify that the MethodImpl entry must use a MethodDef index to refer to a static "body" method.

trylek replied:
Clarified in the latest David's commit.

333fred commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,389 @@ The algorithm is amended as follows:
+(Change first paragraph)
+
+A generic parameter declared on a generic class or generic method can be *constrained* by one or more
+type (for encoding, see *GenericParamConstraint* table in paragraph II.22.21) and by one or more special
```
```suggestion
types (for encoding, see *GenericParamConstraint* table in paragraph II.22.21) and by one or more special
```

333fred replied:
Or `type(s)`

davidwrighton committed:
Update docs/design/specs/Ecma-335-Augments.md

Co-authored-by: Fred Silberberg <fred@silberberg.xyz>

davidwrighton commented on docs/design/specs/Ecma-335-Augments.md:
```diff
@@ -420,3 +420,389 @@ The algorithm is amended as follows:
+
+Edit section 7: The method indexed by *MethodBody* shall be non-virtual if the method indexed
+by MethodDeclaration is static. Otherwise it shall be virtual.
+
```
```suggestion

(Add new section 14 after section 13:)

* 14. If the method indexed by *MethodBody* has the static flag set, the method indexed by *MethodBody* must be indexed via a MethodDef and not a MemberRef. [ERROR]

```

davidwrighton replied:
This is the change that I believe @AlekseyTs wanted to require that MethodImpl records are MethodDef based.

davidwrighton committed:
Update docs/design/specs/Ecma-335-Augments.md
